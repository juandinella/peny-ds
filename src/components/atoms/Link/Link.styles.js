import { css } from 'styled-components'

import layout from '../../../styles/theme/layout'

export default css`
  display: inline-flex;
  color: var(--color-primary-main);
  text-decoration: none;
  transition: color var(--transition-timing-easing);
  cursor: pointer;

  &:hover {
    color: var(--color-primary-dark);
  }

  ${(props) =>
    props.size &&
    css`
      font-size: var(
        --font-${(props) => (props.mSize ? props.mSize : props.size)}
      );

      @media screen and (min-width: ${() => layout.breakpoints.sm}) {
        font-size: var(--font-${(props) => props.size});
      }
    `}
`
