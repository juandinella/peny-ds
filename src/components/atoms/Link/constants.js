export const options = {
  /* Size of the link */
  sizes: ['xxs', 'xs', 'sm', 'md', 'lg', 'xl', 'xxl', 'xxxl'],
}
