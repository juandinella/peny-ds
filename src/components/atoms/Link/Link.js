import React from 'react'
import PropTypes from 'prop-types'
import styles from './Link.styles'
import Skeleton from '../Skeleton'
import Utils from '../../../utils'
import { options } from './constants'

/**
 * Hyperlinks between pages.
 */

const Link = ({
  id,
  href,
  target,
  mSize,
  size,
  children,
  onClick,
  shouldRender,
}) => {
  return (
    <Skeleton
      id={`${id}_skeleton`}
      show={!Utils.emptyValue(shouldRender) && !shouldRender()}
    >
      <a
        css={styles}
        id={id}
        data-testid={id}
        href={href}
        target={target}
        size={size}
        mSize={mSize}
        onClick={onClick || null}
      >
        {children}
      </a>
    </Skeleton>
  )
}

Link.displayName = 'Link'

Link.propTypes = {
  href: PropTypes.string,
  /**
   * Target attribute for a link (i.e. set to _blank to open in new tab)
   * - **_blank** yields a link that opens in a new tab
   * - **_self** yields a link that loads the URL into the same browsing context as the current one. This is the default behavior
   * - **_parent** yields a link that loads the URL into the parent browsing context of the current one. If there is no parent, this behaves the same way as _self
   * - **_top** yields a link that loads the URL into the top-level browsing context. If there is no parent, this behaves the same way as _self.
   */
  target: PropTypes.oneOf(['_blank', '_self', '_parent', '_top']),
  children: PropTypes.node.isRequired,
  /** Link's id */
  id: PropTypes.string,
  /** Link's font size */
  size: PropTypes.oneOf(options.sizes),
  /** Link's font size for Mobile*/
  mSize: PropTypes.oneOf(options.sizes),
  /** Click event */
  onClick: PropTypes.func,
  /**
   * Function that decides if the component should be rendered.
   *
   * @return {boolean} false if component should not be rendered and show skeleton instead. Component renders otherwise.
   */
  shouldRender: PropTypes.func,
}

Link.defaultProps = {
  size: 'sm',
}

export default Link
