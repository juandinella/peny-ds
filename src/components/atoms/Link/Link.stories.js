import Link from './Link'
import { options } from './constants'

import {
  getTemplate,
  getListTemplate,
  getOptionsArgTypes,
} from '../../../helpers/storybook'

const Template = getTemplate(Link)
const ListTemplate = getListTemplate(Link)

export default {
  title: 'Atoms/Link',
  component: Link,
  args: {
    children: 'Im a link',
    href: '#',
  },
  argTypes: {
    size: getOptionsArgTypes(options.sizes),
  },
}

export const Default = Template.bind({})

export const Sizes = ListTemplate.bind({})
Sizes.args = { items: options.sizes.map((size) => ({ size })) }

export const Skeleton = Template.bind({})
Skeleton.args = { shouldRender: () => false }
