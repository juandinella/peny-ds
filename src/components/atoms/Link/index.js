import Link from './Link'

export default Link
export { options } from './constants'
