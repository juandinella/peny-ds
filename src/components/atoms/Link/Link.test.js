import React from 'react'

import { render, fireEvent, cleanup } from '@testing-library/react'

import Link from './Link'

import initStoryshots from '@storybook/addon-storyshots'

initStoryshots({
  storyKindRegex: /^Atoms\/Link$/,
})

const componentProps = {
  id: 'unique_id',
  children: 'Go!',
  onClick: jest.fn(),
}

afterEach(cleanup)
beforeEach(() => {
  componentProps.onClick.mockClear()
})

describe('<Link />', () => {
  test('should render component correctly', () => {
    const { queryByTestId } = render(<Link {...componentProps} />)

    const link = queryByTestId(componentProps.id)

    expect(link).toBeTruthy()
  })

  test('should render skeleton', () => {
    const { queryByTestId } = render(
      <Link {...componentProps} shouldRender={() => false} />
    )

    expect(queryByTestId(`${componentProps.id}_skeleton`)).toBeTruthy()
  })

  test('should render content correctly', () => {
    const { queryByTestId } = render(<Link {...componentProps} />)

    const link = queryByTestId(componentProps.id)

    expect(link.textContent).toEqual(componentProps.children)
  })

  test('should call click function', () => {
    const { container } = render(<Link {...componentProps} />)

    fireEvent.click(container.firstChild)

    expect(componentProps.onClick).toHaveBeenCalledTimes(1)
  })

  test('should have href', () => {
    const { queryByTestId } = render(
      <Link {...componentProps} href="https://google.com/" />
    )

    const link = queryByTestId(componentProps.id)

    expect(link.href).toEqual('https://google.com/')
  })

  test('should have target', () => {
    const { queryByTestId } = render(
      <Link {...componentProps} href="https://google.com/" target="_blank" />
    )

    const link = queryByTestId(componentProps.id)

    expect(link.target).toEqual('_blank')
  })
})
