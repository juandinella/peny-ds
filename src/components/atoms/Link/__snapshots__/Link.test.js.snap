// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/Link Default 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <a
    className="sc-notwuq-0 iDxIsk"
    href="#"
    onClick={null}
    size="sm"
  >
    Im a link
  </a>
</div>
`;

exports[`Storyshots Atoms/Link Sizes 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <a
    className="sc-notwuq-0 fBAtrU"
    href="#"
    onClick={null}
    size="xxs"
  >
    Im a link
  </a>
  <a
    className="sc-notwuq-0 ffVqpg"
    href="#"
    onClick={null}
    size="xs"
  >
    Im a link
  </a>
  <a
    className="sc-notwuq-0 iDxIsk"
    href="#"
    onClick={null}
    size="sm"
  >
    Im a link
  </a>
  <a
    className="sc-notwuq-0 hnMsYk"
    href="#"
    onClick={null}
    size="md"
  >
    Im a link
  </a>
  <a
    className="sc-notwuq-0 kvftMQ"
    href="#"
    onClick={null}
    size="lg"
  >
    Im a link
  </a>
  <a
    className="sc-notwuq-0 jgWTKY"
    href="#"
    onClick={null}
    size="xl"
  >
    Im a link
  </a>
  <a
    className="sc-notwuq-0 bLdJEQ"
    href="#"
    onClick={null}
    size="xxl"
  >
    Im a link
  </a>
  <a
    className="sc-notwuq-0 hOxZvE"
    href="#"
    onClick={null}
    size="xxxl"
  >
    Im a link
  </a>
</div>
`;

exports[`Storyshots Atoms/Link Skeleton 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-1osow0y-0 dpAisn"
    data-testid="undefined_skeleton"
    id="undefined_skeleton"
    style={
      Object {
        "height": undefined,
        "width": undefined,
      }
    }
  >
    <a
      className="sc-notwuq-0 iDxIsk"
      href="#"
      onClick={null}
      size="sm"
    >
      Im a link
    </a>
  </div>
</div>
`;
